staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -417,28 +392,351 @@ func TestPriorityAndFairnessWithPanicRecoverAndTimeoutFilter(t *testing.T) {
+		requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			expectMatchingAPFHeaders(t, w, fsName, plName)
+
+			if r.URL.Path == rquestTimesOutPath {
+				defer close(reqHandlerCompletedCh)
+
+				// inner handler writes header and then let the request time out.
+				w.WriteHeader(http.StatusBadRequest)
+				<-callerRoundTripDoneCh
+
+				// we expect the timeout handler to have timed out this request by now and any attempt
+				// to write to the response should return a http.ErrHandlerTimeout error.
+				if _, err := w.Write([]byte("foo")); err != http.ErrHandlerTimeout {
+					t.Fatalf("expected error: %#v, but got: %#v", http.ErrHandlerTimeout, err)
+				}
+			}
+		})
+		handler := newHandlerChain(t, requestHandler, controller, userName, requestTimeout)
+
+		server, requestGetter := newHTTP2ServerWithClient(handler)
+		defer server.Close()
+
+		var err error
+		func() {
+			defer close(callerRoundTripDoneCh)
+			t.Logf("waiting for the request: %q to time out", rquestTimesOutPath)
+			_, err = requestGetter(rquestTimesOutPath)
+		}()
+
+		t.Logf("waiting for the inner handler of the request: %q to complete", rquestTimesOutPath)
+		<-reqHandlerCompletedCh
+
+		expectResetStreamError(t, err)
+
+		close(stopCh)
+		t.Log("waiting for the controller to shutdown")
+
+		controllerErr := <-controllerCompletedCh
+		if controllerErr != nil {
+			t.Errorf("expected a nil error from controller, but got: %#v", controllerErr)
+		}
+	})
+
+	t.Run("priority level concurrency is set to 1, queue length is 1, enqueued request should time out with 504", func(t *testing.T) {
+		const (
+			requestTimeout                                                     = 3 * time.Second
+			userName                                                           = "alice"
+			fsName                                                             = "test-fs"
+			plName                                                             = "test-pl"
+			serverConcurrency, plConcurrencyShares, plConcurrency, queueLength = 1, 1, 1, 1
+		)
+
+		apfConfiguration := newConfiguration(fsName, plName, userName, plConcurrencyShares, queueLength)
+		stopCh := make(chan struct{})
+		controller, controllerCompletedCh := startAPFController(t, stopCh, apfConfiguration, serverConcurrency, requestTimeout/4, plName, plConcurrency)
+
+		firstRequestTimesOutPath := "/request/first/time-out-as-designed"
+		secondRequestEnqueuedPath := "/request/second/enqueued-as-designed"
+		firstReqHandlerCompletedCh, firstReqInProgressCh := make(chan struct{}), make(chan struct{})
+		firstReqRoundTripDoneCh, secondReqRoundTripDoneCh := make(chan struct{}), make(chan struct{})
+		requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			expectMatchingAPFHeaders(t, w, fsName, plName)
+
+			if r.URL.Path == firstRequestTimesOutPath {
+				defer close(firstReqHandlerCompletedCh)
+
+				close(firstReqInProgressCh)
+				<-firstReqRoundTripDoneCh
```

Comment:
@MikeSpreitzer The second (enqueued) request may be timed out promptly ( within `1/4th` of the request timeout, if we fix p&f in the future)  and the following synchronization issue might happen: 
A: T+0s: first request being executed (timeout = `60s` for example)
B: T+1s: second request arrives and is enqueued (timeout = `60s`)
C: T+16s: second request is rejected with a 429 
D: `secondReqRoundTripDoneCh` unblocks and the first request executes and it does not time out as designed.

The test does not assume when the enqueued request times out, thoughts?

Reply:
Good answer.  In short, we need both waits because the design is to wait for the later of two events and the test deliberately does not make an assumption about which one happens later.